lighthouse-core/test/lib/client-rect-functions-test.js

```diff
@@ -0,0 +1,245 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+const {simplifyClientRects, getRectOverlapArea} = require('../../lib/client-rect-functions');
+const assert = require('assert');
+
+function makeClientRect({x, y, width, height}) {
+  return {
+    left: x,
+    top: y,
+    right: x + width,
+    bottom: y + height,
+    width,
+    height,
+  };
+}
+
+describe('simplifyClientRects', () => {
+  it('Merges rects if a smaller rect is inside a larger one', () => {
+    const res = simplifyClientRects([
+      makeClientRect({
+        x: 10,
+        y: 10,
+        width: 100,
+        height: 10,
+      }),
+      makeClientRect({
+        x: 10,
+        y: 10,
+        width: 50,
+        height: 10,
+      }),
+    ]);
+    assert.deepEqual(res, [
+      makeClientRect({
+        x: 10,
+        y: 10,
+        width: 100,
+        height: 10,
+      }),
+    ]);
+  });
+  it('Merges two horizontally adjacent client rects', () => {
+    const res = simplifyClientRects([
+      makeClientRect({
+        x: 10,
+        y: 10,
+        width: 100,
+        height: 10,
+      }),
+      makeClientRect({
+        x: 110,
+        y: 10,
+        width: 100,
+        height: 10,
+      }),
+    ]);
+    assert.deepEqual(res, [
+      makeClientRect({
+        x: 10,
+        y: 10,
+        width: 200,
+        height: 10,
+      }),
+    ]);
+  });
+
+  it('Merges three horizontally adjacent client rects', () => {
+    const res = simplifyClientRects([
+      makeClientRect({
+        x: 10,
+        y: 10,
+        width: 100,
+        height: 10,
+      }),
+      makeClientRect({
```

Comment:
would be really neat to replace all the x/y/w/h rect definitions with some sort of visual ascii

```
const input = makeRects(`
+--+--+--+
|..|..|..|
+--+--+--+
`);
const simple = simplifyClientRects(input)

assertEqual(simple, makeRect(`
+--------+
|........|
+--------+
`));
```

Obviously this is way more work. Maybe if we had dozens of these tests it'd be worth it :)